exonum/src/blockchain/service/srv.rs

```diff
@@ -0,0 +1,247 @@
+extern crate core;
+
+use serde_json::Value;
+
+use iron::prelude::*;
+use iron::Handler;
+use router::Router;
+
+use serde_json;
+
+use storage::{Fork, Snapshot};
+use messages::RawTransaction;
+use encoding::Error as MessageError;
+use crypto::Hash;
+use api::{ApiError, Api};
+use node::{TransactionSend, ApiSender};
+
+
+use super::{Transaction, ServiceContext, ApiContext, Service};
+use std::marker::PhantomData;
+
+/// TODO
+pub trait TransactionSet {
+    /// TODO
+    fn tx_from_raw(raw: RawTransaction) -> Result<Box<Transaction>, MessageError>;
+    /// TODO
+    fn tx_from_request(request: &mut Request) -> Result<Box<Transaction>, ApiError>;
+}
+
+#[macro_export]
+macro_rules! transaction_set {
+    ( $name:ident { $($tx:ident),* } ) => {
+        pub enum $name {
+        }
+
+        impl $crate::blockchain::TransactionSet for $name {
+            fn tx_from_raw(raw: RawTransaction) -> Result<Box<$crate::blockchain::Transaction>, $crate::encoding::Error> {
+                let message_type = raw.message_type();
+                $(
+                    if $tx::message_id() == message_type {
```

Comment:
Could it make sense to use `match` instead of `if`s here? IMO, it would read more naturally (not sure about performance, but intuitively, `match` leaves more room for optimizations).

Reply:
`match` does not work here because `message_id` is a function. Match would work though with associated constants, and I'd love to change message_id to such constant.